### PR TITLE
[MM-16934] Add API endpoint to update channel privacy

### PR DESCRIPTION
**Pull request by streamer45**
#### Summary

This PR adds a `/channels/privacy` API endpoint to make it possible to change the privacy setting of a channel from public to private and back.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-16934

### CODE REVIEW

File: api4/channel_test.go [hunk 1]
```diff
@@ -1387,6 +1387,55 @@ func TestConvertChannelToPrivate(t *testing.T) {
...
+	CheckNoError(t, resp)
+
+	_, resp = Client.UpdateChannelPrivacy(privateChannel.Id, model.CHANNEL_PRIVATE)
+	CheckNoError(t, resp)
+
+	updatedChannel, resp := Client.UpdateChannelPrivacy(publicChannel.Id, model.CHANNEL_PRIVATE)
+	CheckNoError(t, resp)
+
+	if updatedChannel.Type != model.CHANNEL_PRIVATE {
+		t.Fatal("public channel should update privacy to private")
```
Comment (jespino) [inline api4/channel_test.go]: Use `assert.Equal` instead of `t.Fatal` approach.

File: api4/channel_test.go [hunk 2]
```diff
@@ -1387,6 +1387,55 @@ func TestConvertChannelToPrivate(t *testing.T) {
...
+	}
+
+	updatedChannel, resp = Client.UpdateChannelPrivacy(privateChannel.Id, model.CHANNEL_OPEN)
+	CheckNoError(t, resp)
+
+	if updatedChannel.Type != model.CHANNEL_OPEN {
+		t.Fatal("private channel should update privacy to public")
```
Comment (jespino) [inline api4/channel_test.go]: Same here, assert instead of `t.Fatal`

File: api4/channel_test.go (hunk 1 above)
Comment (jespino) [reply api4/channel_test.go]: Another thing, to be more "sure" can we "get again" the channel from the database? In addition to check the returned value, check the result of getting again the channel.

File: api4/channel_test.go (hunk 2 above)
Comment (jespino) [reply api4/channel_test.go]: And same here about getting again the channel.

File: api4/channel_test.go [hunk 3]
```diff
@@ -1387,6 +1387,67 @@ func TestConvertChannelToPrivate(t *testing.T) {
 	}
 }
 
+func TestUpdateChannelPrivacy(t *testing.T) {
```
Comment (jespino) [inline api4/channel_test.go]: Not something needed, but we are pretty big fans of use table testing when is possible and makes sense. Here looks like a table testing can be a good option, but up to you :)

Comment (streamer45) [reply api4/channel_test.go]: Sounds like a great idea. I pushed an update.

Comment (jespino) [reply api4/channel_test.go]: Nice

**Comment by streamer45** [comment]
> This looks fine, there's just one thing that feels a bit uncomfortable to me: we have the `channel/foo/convert` API endpoint to make a channel private, and then `channel/foo/privacy` to make it public. Feels a bit weird. Could we have one for both, or a more consistent naming scheme, or is there a reason we went for this?

I share the same feeling. Originally I wanted to make `/channels/{channel_id}/convert` also work in the other direction but figured out that doing so would create a non backward compatible change unless we used something pretty ugly like query params to cover the additional case.

After discussing it with @jespino I thought it was best to create a new API endpoint that would cover both cases. `/channels/{channel_id}/privacy` will go in both direction (private to public and public to private) and could effectively replace all instances of `/convert'.

I am open to suggestion as far as naming but I don't dislike "privacy" as opposed to "convert" which I personally find more ambiguous and less extensible.

**Pull request merged**